cob_obstacle_distance/src/distance_manager.cpp

```diff
@@ -0,0 +1,386 @@
+    {
+        ROS_ERROR("Failed to get parameter \"chain_tip_link\".");
+        return -4;
+    }
+
+    if(!nh_.getParam("joint_names", this->joints_))
+    {
+        ROS_ERROR("Failed to get parameter \"joint_names\".");
+        return -5;
+    }
+
+    robot_structure.getChain(this->chain_base_link_, this->chain_tip_link_, this->chain_);
+    if(chain_.getNrOfJoints() == 0)
+    {
+        ROS_ERROR("Failed to initialize kinematic chain");
+        return -6;
+    }
+
+    for(uint16_t i = 0; i < chain_.getNrOfSegments(); ++i)
+    {
+        KDL::Segment s = chain_.getSegment(i);
+        this->segments_.push_back(s.getName());
+        ROS_INFO_STREAM("Managing Segment Name: " << s.getName());
+    }
+
+    this->marker_pub_ = this->nh_.advertise<visualization_msgs::Marker>("visualization_marker", 1);
+    this->obstacle_distances_pub_ = this->nh_.advertise<cob_obstacle_distance::ObstacleDistances>(this->robo_namespace_, 1);
+
+    obstacle_mgr_.reset(new ShapesManager(this->marker_pub_));
+    object_of_interest_mgr_.reset(new ShapesManager(this->marker_pub_));
+
+    adv_chn_fk_solver_pos_.reset(new AdvancedChainFkSolverPos_recursive(chain_));
+
+    last_q_ = KDL::JntArray(chain_.getNrOfJoints());
+    last_q_dot_ = KDL::JntArray(chain_.getNrOfJoints());
+
+    jnt2jac_.reset(new KDL::ChainJntToJacSolver(chain_));
+    return 0;
+}
+
+void DistanceManager::clear()
+{
+    this->obstacle_mgr_->clear();
+    this->object_of_interest_mgr_->clear();
+}
+
+bool DistanceManager::waitForMarkerSubscriber()
+{
+    while (this->marker_pub_.getNumSubscribers() < 1)
+    {
+      if (!ros::ok())
+      {
+        return false;
+      }
+
+      ROS_WARN_ONCE("Please create a subscriber to the marker");
+      sleep(1);
+    }
+
+    return true;
+}
+
+void DistanceManager::addObstacle(tPtrMarkerShapeBase s)
+{
+    this->obstacle_mgr_->addShape(s);
+}
+
+void DistanceManager::addObjectOfInterest(tPtrMarkerShapeBase s)
+{
+    this->object_of_interest_mgr_->addShape(s);
+}
+
+void DistanceManager::drawObstacles(bool enforceDraw)
+{
+    this->obstacle_mgr_->draw(enforceDraw);
+}
+
+void DistanceManager::drawObjectsOfInterest(bool enforceDraw)
+{
+    this->object_of_interest_mgr_->draw(enforceDraw);
+}
+
+bool DistanceManager::collide(tPtrMarkerShapeBase s1, tPtrMarkerShapeBase s2)
+{
+    fcl::CollisionObject x = s1->getCollisionObject();
+    fcl::CollisionObject y = s2->getCollisionObject();
+    fcl::CollisionResult result;
+    fcl::CollisionRequest request(1, true);
+    fcl::collide(&x, &y, request, result);
+    ROS_INFO_STREAM("isCollision: " << result.isCollision() << std::endl);
+}
+
+void DistanceManager::calculate()
+{
+    bool initial = true;
+    cob_obstacle_distance::ObstacleDistances obstacle_distances;
+    for(t_map_ObstacleDistance_iter it = this->obstacle_distances_.begin(); it != this->obstacle_distances_.end(); ++it)
+    {
+        if(initial)
+        {
+            if (!this->transform())
```

Comment:
When calculate() is called then transform() is called but only in the case when there are registrations. If there are no registrations then it is not necessary to call transform() all the time. 

The loop iterates over all managed reservations for each calculate() call. 

So this will be sufficient either for the case with base_active as well as the case without it.